storage/src/index/ordered.rs

```diff
@@ -0,0 +1,388 @@
+
+                // Remove anything that is prunable.
+                prune_with_cursor(&mut cursor, predicate);
+            }
+            BTreeEntry::Vacant(_) => {}
+        }
+    }
+
+    fn remove(&mut self, key: &[u8]) {
+        // To ensure metrics are accurate, we iterate over all conflicting values and remove them
+        // one-by-one (rather than just removing the entire entry).
+        self.prune(key, |_| true);
+    }
+
+    #[cfg(test)]
+    fn items(&self) -> usize {
+        self.items.get() as usize
+    }
+
+    #[cfg(test)]
+    fn pruned(&self) -> usize {
+        self.pruned.get() as usize
+    }
+}
+
+impl<T: Translator, V: Eq> Drop for Index<T, V> {
+    /// To avoid stack overflow on keys with many collisions, we implement an iterative drop (in
+    /// lieu of Rust's default recursive drop).
+    fn drop(&mut self) {
+        for (_, record) in self.map.iter_mut() {
+            let mut next = record.next.take();
+            while let Some(mut record) = next {
+                next = record.next.take();
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::translator::OneCap;
+    use commonware_macros::test_traced;
+    use commonware_runtime::{deterministic, Runner};
+
+    #[test_traced]
+    fn test_ordered_index_ordering() {
+        let runner = deterministic::Runner::default();
+        runner.start(|context| async move {
+            let mut index = Index::<_, u64>::init(context, OneCap);
+            assert_eq!(index.keys(), 0);
+
+            let k1 = &[0x0b, 0x02, 0xAA]; // translated key 0b
+            let k2 = &[0x1c, 0x04, 0xCC]; // translated key 1c
+            let k2_collides = &[0x1c, 0xBB, 0x11];
+            let k3 = &[0x2d, 0x06, 0xEE]; // translated key 2d
+            index.insert(k1, 1);
+            index.insert(k2, 21);
+            index.insert(k2_collides, 22);
+            index.insert(k3, 3);
+            assert_eq!(index.keys(), 3);
+
+            // First translated key is 0b.
+            let mut next = index.first_translated_key_values();
+            assert_eq!(next.next().unwrap(), &1);
+            assert_eq!(next.next(), None);
+
+            // Next translated key to 0x00 is 0b.
+            let mut next = index.next_translated_key_values(&[0x00]);
+            assert_eq!(next.next().unwrap(), &1);
+            assert_eq!(next.next(), None);
+
+            // Next translated key to 0x0b is 1c.
+            let mut next = index.next_translated_key_values(&[0x0b, 0x01, 0x02]);
+            assert_eq!(next.next().unwrap(), &21);
+            assert_eq!(next.next().unwrap(), &22);
+            assert_eq!(next.next(), None);
+
+            // Next translated key to 0x1b is 1c.
+            let mut next = index.next_translated_key_values(&[0x1b, 0x01, 0x02, 0x03]);
+            assert_eq!(next.next().unwrap(), &21);
+            assert_eq!(next.next().unwrap(), &22);
```

Comment:
For clarity, I recommend setting the k2_collides key (which is after k2) to have a value lower than k2.

Reply:
done